src/main/scala/org/broadinstitute/dsde/agora/server/webservice/routes/OpenIdConnectDirectives.scala

```diff
@@ -2,25 +2,32 @@ package org.broadinstitute.dsde.agora.server.webservice.routes
 
 import akka.http.scaladsl.server.Directive1
 import akka.http.scaladsl.server.Directives._
+import org.broadinstitute.dsde.agora.server.dataaccess.SamClient
 import org.broadinstitute.dsde.agora.server.webservice.util.ImplicitMagnet
 
 import scala.concurrent.ExecutionContext
 
 
 /**
  * Directives class to retrieve user information from the OpenId Connect proxy
- * provided information in the header
+ * provided information in the header.
  */
 class OpenIdConnectDirectives extends AgoraDirectives {
+  private val serviceAccountDomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r
 
-  override def commonNameFromRequest(magnet: ImplicitMagnet[ExecutionContext]): Directive1[String] = {
-    headerValueByName("OIDC_CLAIM_name")
+  private def isServiceAccount(email: String) = {
+    serviceAccountDomain.pattern.matcher(email).matches
   }
 
   override def usernameFromRequest(magnet: ImplicitMagnet[ExecutionContext]): Directive1[String] = {
-    headerValueByName("OIDC_CLAIM_email")
+    (headerValueByName("OIDC_CLAIM_email") & headerValueByName("OIDC_access_token")).tmap { case (email, token) =>
+      if (isServiceAccount(email)) {
+        SamClient.getUserEmail(token).getOrElse(email)
```

Comment:
This is making a synchronous HTTP call, which is something new in this flow. `usernameFromRequest` is called pretty frequently, from many (all? almost all?) endpoints. Is it okay that this is synchronous? Are there performance considerations we should think about? I don't know what the volume of requests to Agora is. Should we consider caching the result of the call to Sam?

Reply:
I don't want to cache until we have data to suggest there is a problem.

I agree with the async comment and made it so since this is within the routing.

Reply:
Do we have some confidence that it won't be an immediate problem or do we need to gather some data before putting this change into production? Do we know how frequently these endpoints are called, such as based on user activity browsing method repo, or based on the number of submissions, or based on the number of individual jobs? This could affect more than just Agora; it's asking Sam to do more work.

Reply:
Over the highest volume 3 hour period in the last 7 days this api saw 23K requests. 13K responded in 15ms or less and almost all were under 100ms. The request rate from the rawls side is also throttled as there are only 10 workers responsible for submitting workflows to cromwell.

Reply:
it looks like the highest sustained workflow submission rate to cromwell over the past 90 days was about 90/minute